File: backend/scraper/lacentrale.py
"""LaCentrale listing scraper.

LaCentrale is a SPA fronted by Datadome. Instead of guessing CSS
selectors we intercept the JSON API responses the frontend fires when
loading a search, and parse listings from those payloads. This is the
same approach that worked for Leboncoin.

If Datadome shows a challenge, run with `--debug` — the browser opens
visibly so you can solve it by hand, and the search will load afterwards.
"""

from __future__ import annotations

import asyncio
import json
import random
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any
from urllib.parse import urljoin

from playwright.async_api import Browser, Response, async_playwright

BASE_URL = "https://www.lacentrale.fr"
SEARCH_URL = (
    f"{BASE_URL}/listing"
    "?makesModelsCommercialNames=TESLA%3A%3AMODEL%203"
    "&sortBy=priceAsc"
)

USER_AGENT = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
    "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/147.0.0.0 Safari/537.36"
)

DEBUG_DIR = Path(__file__).resolve().parent.parent / "debug"


@dataclass
class ScrapedListing:
    source: str
    external_id: str
    title: str
    make: str | None
    model: str | None
    version: str | None
    price_eur: int | None
    year: int | None
    mileage_km: int | None
    fuel: str | None
    gearbox: str | None
    location: str | None
    url: str
    image_url: str | None


def _parse_int(value: Any) -> int | None:
    if value is None:
        return None
    digits = re.sub(r"[^\d]", "", str(value))
    return int(digits) if digits else None


def _first(node: dict, *keys: str) -> Any:
    for key in keys:
        if key in node and node[key] not in (None, ""):
            return node[key]
    return None


def _find_listings(payload: Any) -> list[dict]:
    """Walk the JSON tree looking for an array of listing-like dicts.

    LaCentrale has changed the nesting of search results several times
    (hits, classifieds, results, vehicles, items...). A recursive search
    for arrays of objects with listing-y keys is more resilient than
    hard-coding a path.
    """
    listing_keys = {"classified_id", "classifiedId", "adId", "ad_id", "id"}
    title_keys = {"title", "commercialName", "model", "make", "brand"}
    found: list[dict] = []
    seen_ids: set[str] = set()

    def looks_like_listing(obj: dict) -> bool:
        keys = set(obj.keys())
        if not keys & listing_keys:
            return False
        if not keys & title_keys:
            return False
        # A listing should have at least one price-ish or vehicle-ish field.
        vehicle_markers = {"price", "customerPrice", "mileage", "km", "year", "energy", "fuel"}
        return bool(keys & vehicle_markers)

    def walk(node: Any):
        if isinstance(node, dict):
            if looks_like_listing(node):
                ident = str(_first(node, "classified_id", "classifiedId", "adId", "ad_id", "id"))
                if ident and ident not in seen_ids:
                    seen_ids.add(ident)
                    found.append(node)
                return
            for v in node.values():
                walk(v)
        elif isinstance(node, list):
            for v in node:
                walk(v)

    walk(payload)
    return found


def _extract_image_url(ad: dict) -> str | None:
    for key in ("mainPhoto", "photo", "thumbnail", "image"):
        value = ad.get(key)
        if isinstance(value, str) and value.startswith("http"):
            return value
        if isinstance(value, dict):
            for sub in ("url", "src", "large", "medium", "small"):
                if isinstance(value.get(sub), str):
                    return value[sub]

    photos = ad.get("photos") or ad.get("thumbnails") or ad.get("images")
    if isinstance(photos, list) and photos:
        first = photos[0]
        if isinstance(first, str):
            return first
        if isinstance(first, dict):
            for sub in ("url", "src", "large", "medium", "small"):
                if isinstance(first.get(sub), str):
                    return first[sub]
    return None


def _ad_to_listing(ad: dict) -> ScrapedListing | None:
    ident = _first(ad, "classified_id", "classifiedId", "adId", "ad_id", "id")
    if ident is None:
        return None

    make = _first(ad, "make", "brand")
    model = _first(ad, "model")
    version = _first(ad, "version", "commercialName", "trim")
    title_parts = [str(p) for p in (make, model, version) if p]
    title = _first(ad, "title") or " ".join(title_parts)

    price_eur = _parse_int(_first(ad, "customerPrice", "price", "priceValue"))
    year = _parse_int(_first(ad, "year", "firstRegistrationYear"))
    if year is None:
        reg = _first(ad, "registration_date", "firstRegistrationDate", "registrationDate")
        if isinstance(reg, str):
            m = re.search(r"(19|20)\d{2}", reg)
            if m:
                year = int(m.group(0))

    mileage_km = _parse_int(_first(ad, "mileage", "km", "kilometers"))
    fuel = _first(ad, "energy", "fuel", "energyLabel")
    gearbox = _first(ad, "gearbox", "transmission")

    city = _first(ad, "city", "cityLabel")
    zipcode = _first(ad, "zipCode", "zipcode", "postalCode")
    location = " ".join(str(p) for p in (city, zipcode) if p) or None

    detail_url = _first(ad, "urlDetail", "url", "link", "detailUrl")
    if isinstance(detail_url, str):
        url = urljoin(BASE_URL, detail_url)
    else:
        url = f"{BASE_URL}/auto-occasion-annonce-{ident}.html"

    image_url = _extract_image_url(ad)

    return ScrapedListing(
        source="lacentrale",
        external_id=str(ident),
        title=title.strip() if title else str(ident),
        make=str(make).title() if make else None,
        model=str(model).title() if model else None,
        version=str(version) if version else None,
        price_eur=price_eur,
        year=year,
        mileage_km=mileage_km,
        fuel=str(fuel) if fuel else None,
        gearbox=str(gearbox) if gearbox else None,
        location=location,
        url=url,
        image_url=image_url,
    )


async def _scrape_page(browser: Browser, url: str, debug: bool = False) -> list[ScrapedListing]:
    context = await browser.new_context(
        user_agent=USER_AGENT,
        viewport={"width": 1440, "height": 900},
        locale="fr-FR",
    )
    page = await context.new_page()

    captured_payloads: list[tuple[str, dict]] = []

    async def on_response(response: Response) -> None:
        try:
            req_url = response.url
            if "lacentrale" not in req_url:
                return
            ctype = (response.headers or {}).get("content-type", "")
            if "json" not in ctype:
                return
            # Search / listing API endpoints: keep the filter loose.
            if not any(k in req_url for k in ("search", "listing", "classified", "/api/", "graphql")):
                return
            try:
                data = await response.json()
            except Exception:
                return
            if isinstance(data, (dict, list)):
                captured_payloads.append((req_url, data))
        except Exception:
            pass

    page.on("response", lambda r: asyncio.create_task(on_response(r)))

    try:
        await page.goto(url, wait_until="domcontentloaded", timeout=30_000)
        await page.wait_for_timeout(random.randint(1500, 3000))
        try:
            await page.get_by_role("button", name=re.compile("accepter|accept|tout accepter", re.I)).click(timeout=2000)
        except Exception:
            pass
        try:
            await page.wait_for_load_state("networkidle", timeout=12_000)
        except Exception:
            pass
        await page.wait_for_timeout(random.randint(1000, 2000))

        if debug:
            DEBUG_DIR.mkdir(exist_ok=True)
            await page.screenshot(path=str(DEBUG_DIR / "lacentrale.png"), full_page=True)
            (DEBUG_DIR / "lacentrale.html").write_text(await page.content(), encoding="utf-8")
            for i, (req_url, payload) in enumerate(captured_payloads):
                (DEBUG_DIR / f"lacentrale_api_{i:02d}.json").write_text(
                    f"// {req_url}\n" + json.dumps(payload, ensure_ascii=False, indent=2),
                    encoding="utf-8",
                )
            print(
                f"  [debug] saved screenshot, html, "
                f"{len(captured_payloads)} api payload(s) to {DEBUG_DIR}"
            )

        ads: list[dict] = []
        for _, payload in captured_payloads:
            ads.extend(_find_listings(payload))

        # Dedupe by external id (same ad can appear across multiple API calls).
        unique: dict[str, dict] = {}
        for ad in ads:
            ident = _first(ad, "classified_id", "classifiedId", "adId", "ad_id", "id")
            if ident is not None:
                unique.setdefault(str(ident), ad)

        listings: list[ScrapedListing] = []
        for ad in unique.values():
            try:
                listing = _ad_to_listing(ad)
                if listing:
                    listings.append(listing)
            except Exception as exc:
                print(f"  ! ad parse failed: {exc}")
        return listings
    finally:
        await context.close()


async def scrape(search_url: str = SEARCH_URL, pages: int = 1, debug: bool = False) -> list[ScrapedListing]:
    all_results: list[ScrapedListing] = []
    async with async_playwright() as pw:
        browser = await pw.chromium.launch(headless=not debug)
        try:
            for page_num in range(1, pages + 1):
                if page_num == 1:
                    url = search_url
                else:
                    sep = "&" if "?" in search_url else "?"
                    url = f"{search_url}{sep}page={page_num}"
                print(f"[lacentrale] page {page_num}: {url}")
                results = await _scrape_page(browser, url, debug=debug)
                print(f"  -> {len(results)} listings")
                all_results.extend(results)
                if page_num < pages:
                    await asyncio.sleep(random.uniform(3, 6))
        finally:
            await browser.close()
    return all_results


if __name__ == "__main__":
    listings = asyncio.run(scrape(pages=1, debug=True))
    for listing in listings[:5]:
        print(listing)
    print(f"Total: {len(listings)}")

File: backend/scraper/lacentrale.js
'use strict'

const fs = require('fs')
const os = require('os')
const path = require('path')
const { chromium } = require('playwright-extra')
const StealthPlugin = require('puppeteer-extra-plugin-stealth')

chromium.use(StealthPlugin())

const BASE_URL = 'https://www.lacentrale.fr'
const SEARCH_URL = `${BASE_URL}/listing?makesModelsCommercialNames=TESLA&type=vo`
const LOGIN_URL = `${BASE_URL}/login`
const PROFILE_DIR = path.join(os.homedir(), '.teslapricing', 'lacentrale-profile')

const { USER_AGENT } = require('./constants')

const LAUNCH_ARGS = [
  '--no-sandbox',
  '--disable-dev-shm-usage',
  '--disable-infobars',
  '--window-size=1440,900',
]

function canonicalModel(text) {
  if (!text) return null
  if (/model[\s_-]*x/i.test(text)) return 'Model X'
  if (/model[\s_-]*s/i.test(text)) return 'Model S'
  if (/model[\s_-]*y/i.test(text)) return 'Model Y'
  if (/model[\s_-]*3/i.test(text)) return 'Model 3'
  return null
}

// Extracts listing cards directly from the rendered DOM via page.evaluate()
async function extractFromDom(page) {
  const raw = await page.evaluate((baseUrl) => {
    const cards = document.querySelectorAll('[data-tracking-meta]')
    const results = []

    for (const card of cards) {
      try {
        let ref = null
        try { ref = JSON.parse(card.getAttribute('data-tracking-meta')).classified_ref } catch {}
        if (!ref) continue

        const anchor = card.querySelector('a[href*="auto-occasion"]')
        const href = anchor ? anchor.getAttribute('href') : null
        const url = href ? (href.startsWith('http') ? href : baseUrl + href) : null

        const title = card.querySelector('h2')?.textContent?.trim() || null
        // subtitle is the first non-h2 text block after the title (version/trim)
        const subtitleEl = card.querySelector('[class*="subTitle"]')
        const version = subtitleEl?.textContent?.trim() || null

        // Characteristic items: year, gearbox, km, fuel — identify by content pattern
        const chars = [...card.querySelectorAll('[class*="vehicleCharacteristicsItem"]')]
          .map(el => el.querySelector('[class*="Text_Text_body-medium"]')?.textContent?.trim())
          .filter(Boolean)

        let year = null, mileage_km = null, fuel = null, gearbox = null
        let color = null, horse_power = null, doors = null, seats = null
        for (const c of chars) {
          if (/^\d{4}$/.test(c)) { year = parseInt(c, 10); continue }
          if (/km/i.test(c)) { mileage_km = parseInt(c.replace(/[^\d]/g, ''), 10) || null; continue }
          if (/électrique|diesel|essence|hybride|gpl|hydrogène/i.test(c)) { fuel = c; continue }
          if (/auto|manuelle|automatique/i.test(c)) { gearbox = c; continue }
          if (/\bch\b|\bcv\b/i.test(c)) { const n = parseInt(c.replace(/[^\d]/g, ''), 10); if (n) horse_power = n; continue }
          if (/portes?/i.test(c)) { const n = parseInt(c.replace(/[^\d]/g, ''), 10); if (n) doors = n; continue }
          if (/places?/i.test(c)) { const n = parseInt(c.replace(/[^\d]/g, ''), 10); if (n) seats = n; continue }
          if (/^(?:gris|blanc|noir|rouge|bleu|vert|orange|jaune|marron|beige|argent|violet|rose|bordeaux|anthracite|grise|bleue|verte|blanche|noire|rouge|dorée?|dorée?|gold|silver|white|black|blue|red|green|grey|gray|brown)/i.test(c)) { color = c; continue }
        }

        // Price: text inside the price container, strip non-numeric except space
        const priceEl = card.querySelector('[class*="vehiclePrice"]')
        const priceText = priceEl?.textContent?.replace(/[^\d]/g, '') || ''
        const price_eur = priceText ? parseInt(priceText, 10) : null

        // Image: first img with a lacentrale pictures CDN src
        const img = card.querySelector('img[src*="pictures.lacentrale"]')
        const image_url = img?.getAttribute('src') || null

        // Location: text in footer area containing a city/region
        const locationEl = card.querySelector('[class*="sellerLocation"], [class*="location"], [class*="city"]')
        const location = locationEl?.textContent?.trim() || null

        results.push({ ref, url, title, version, year, mileage_km, fuel, gearbox, price_eur, image_url, location, color, horse_power, doors, seats })
      } catch {}
    }
    return results
  }, BASE_URL)

  const seen = new Set()
  const listings = []
  for (const r of raw) {
    if (!r.ref || seen.has(r.ref)) continue
    seen.add(r.ref)

    const titleText = r.title || `Tesla ${r.ref}`
    const make = 'Tesla'
    const model = canonicalModel(titleText)

    listings.push({
      source: 'lacentrale',
      external_id: r.ref,
      title: titleText,
      make,
      model,
      version: r.version || null,
      price_eur: r.price_eur && r.price_eur > 1000 ? r.price_eur : null,
      year: r.year || null,
      mileage_km: r.mileage_km || null,
      fuel: r.fuel || null,
      gearbox: r.gearbox || null,
      color: r.color || null,
      horse_power: r.horse_power || null,
      doors: r.doors || null,
      seats: r.seats || null,
      location: r.location || null,
      url: r.url || `${BASE_URL}/auto-occasion-annonce-${r.ref}.html`,
      image_url: r.image_url || null,
      _photos: [],
    })
  }
  return listings
}

async function isCaptchaPage(page) {
  const title = await page.title()
  if (title === 'lacentrale.fr') return true
  try {
    const body = await page.locator('body').textContent({ timeout: 2000 })
    return /verification required|slide right|datadome|captcha/i.test(body)
  } catch {
    return false
  }
}

async function scrape({ pages = 1, headed = false, debug = false, onPage } = {}) {
  fs.mkdirSync(PROFILE_DIR, { recursive: true })

  const context = await chromium.launchPersistentContext(PROFILE_DIR, {
    headless: !headed,
    args: LAUNCH_ARGS,
    slowMo: 150,
    userAgent: USER_AGENT,
    viewport: { width: 1440, height: 900 },
    locale: 'fr-FR',
  })

  const page = await context.newPage()
  const all = new Map()

  try {
    for (let p = 1; p <= pages; p++) {
      const url = p === 1 ? SEARCH_URL : `${SEARCH_URL}&page=${p}`
      console.log(`[lacentrale] page ${p}: ${url}`)

      await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 30000 })
      await page.waitForTimeout(1500 + Math.random() * 1500)

      if (p === 1) {
        try {
          await page.getByRole('button', { name: /accepter|accept/i }).click({ timeout: 2000 })
        } catch {}
      }

      try { await page.waitForLoadState('networkidle', { timeout: 10000 }) } catch {}
      await page.waitForTimeout(1000 + Math.random() * 1000)

      if (await isCaptchaPage(page)) {
        if (headed) {
          console.log('  ! Captcha detected. Solve it in the browser window, then press Enter here…')
          await new Promise(resolve => process.stdin.once('data', resolve))
          await page.reload({ waitUntil: 'networkidle', timeout: 30000 }).catch(() => {})
          await page.waitForTimeout(1500)
          if (await isCaptchaPage(page)) { console.error('  ! Still blocked. Aborting.'); break }
        } else {
          console.error('  ! Captcha detected. Run once in headed mode to solve it:')
          console.error('  !   node scraper/cli.js lacentrale --headed')
          break
        }
      }

      if (debug) {
        const debugDir = path.join(os.homedir(), '.teslapricing', 'debug')
        fs.mkdirSync(debugDir, { recursive: true })
        const pageHtml = await page.content()
        fs.writeFileSync(path.join(debugDir, `lacentrale-page${p}.html`), pageHtml)
        console.log(`  [debug] HTML → ${path.join(debugDir, `lacentrale-page${p}.html`)}`)
      }

      const listings = await extractFromDom(page)
      console.log(`  -> ${listings.length} listings`)
      const pageListings = listings.filter(l => !all.has(l.external_id))
      for (const l of listings) all.set(l.external_id, l)
      if (onPage && pageListings.length > 0) await onPage(pageListings)

      if (p < pages) await new Promise(r => setTimeout(r, 3000 + Math.random() * 3000))
    }
  } finally {
    await page.close()
    await context.close()
  }
  return [...all.values()]
}

async function login() {
  fs.mkdirSync(PROFILE_DIR, { recursive: true })

  const context = await chromium.launchPersistentContext(PROFILE_DIR, {
    headless: false,
    args: LAUNCH_ARGS,
    slowMo: 150,
    userAgent: USER_AGENT,
    viewport: { width: 1440, height: 900 },
    locale: 'fr-FR',
  })

  const page = await context.newPage()
  await page.goto(LOGIN_URL, { waitUntil: 'domcontentloaded', timeout: 30000 })
  console.log('[lacentrale] Browser open — log in to your account, then press Enter here…')
  await new Promise(resolve => process.stdin.once('data', resolve))
  await context.close()
  console.log('[lacentrale] Session saved. Future scrapes will use your account.')
}

module.exports = { scrape, login }
